File: src/main/java/com/github/nsc/de/compiler/parser/node/VariableType.java
package com.github.nsc.de.compiler.parser.node;

public class VariableType {

    public static VariableType DYNAMIC = new VariableType(Type.DYNAMIC);
    public static VariableType BYTE = new VariableType(Type.BYTE);
    public static VariableType SHORT = new VariableType(Type.SHORT);
    public static VariableType INTEGER = new VariableType(Type.INTEGER);
    public static VariableType LONG = new VariableType(Type.LONG);
    public static VariableType FLOAT = new VariableType(Type.FLOAT);
    public static VariableType DOUBLE = new VariableType(Type.DOUBLE);
    public static VariableType BOOLEAN = new VariableType(Type.BOOLEAN);
    public static VariableType CHAR = new VariableType(Type.CHAR);
    public static VariableType ARRAY = new VariableType(Type.ARRAY);
    public static VariableType OBJECT = new VariableType(Type.OBJECT);

    private final Type type;
    private final String subtype;

    public VariableType(Type type, String subtype) {
        this.type = type;
        this.subtype = subtype;
    }

    public VariableType(Type type) {
        this.type = type;
        this.subtype = null;
    }

    public Type getType() {
        return type;
    }

    public String getSubtype() {
        return subtype;
    }

    public enum Type {
        DYNAMIC,
        BYTE,
        SHORT,
        INTEGER,
        LONG,
        FLOAT,
        DOUBLE,
        BOOLEAN,
        CHAR,
        ARRAY,
        OBJECT,
    }
}

File: shake/compiler/parser/src/commonMain/kotlin/io/github/shakelang/shake/parser/node/VariableType.kt
package io.github.shakelang.shake.parser.node

import io.github.shakelang.shason.JSON
import kotlin.jvm.JvmField

@Suppress("unused")
class VariableType {
    val type: Type
    val subtype: IdentifierNode?

    constructor(subtype: IdentifierNode?) {
        type = Type.OBJECT
        this.subtype = subtype
    }

    constructor(type: Type) {
        this.type = type
        subtype = null
    }

    override fun toString(): String = JSON.stringify(this.json)

    val json: Map<String, *>
        get() = toJson()

    fun toJson(): Map<String, *> =
        mapOf(
            "name" to "VariableType",
            "type" to type,
            "subtype" to subtype,
        )

    enum class Type {
        DYNAMIC, BYTE, SHORT, INTEGER, LONG, FLOAT, DOUBLE, BOOLEAN, CHAR, ARRAY, OBJECT, VOID;

        override fun toString(): String {
            return name.lowercase()
        }
    }

    companion object {
        @JvmField
        val DYNAMIC = VariableType(Type.DYNAMIC)
        @JvmField
        val BYTE = VariableType(Type.BYTE)
        @JvmField
        val SHORT = VariableType(Type.SHORT)
        @JvmField
        val INTEGER = VariableType(Type.INTEGER)
        @JvmField
        val LONG = VariableType(Type.LONG)
        @JvmField
        val FLOAT = VariableType(Type.FLOAT)
        @JvmField
        val DOUBLE = VariableType(Type.DOUBLE)
        @JvmField
        val BOOLEAN = VariableType(Type.BOOLEAN)
        @JvmField
        val CHAR = VariableType(Type.CHAR)
        @JvmField
        val ARRAY = VariableType(Type.ARRAY)
        @JvmField
        val OBJECT = VariableType(Type.OBJECT)
        @JvmField
        val VOID = VariableType(Type.VOID)
    }
}